sqld/src/admin_api.rs

```diff
@@ -85,23 +90,52 @@ async fn handle_post_block<F: MakeNamespace>(
     }
 }
 
-async fn handle_create_namespace_with_dump<F: MakeNamespace>(
+async fn handle_create_namespace<F: MakeNamespace>(
     State(app_state): State<Arc<AppState<F>>>,
     Path(namespace): Path<String>,
-    body: BodyStream,
+    Json(req): Json<CreateNamespaceReq>,
 ) -> Result<(), crate::error::Error> {
-    let dump = Box::new(body.map_err(|e| std::io::Error::new(ErrorKind::Other, e)));
+    let maybe_dump: Option<DumpStream> =
+        OptionFuture::from(req.dump_url.as_ref().map(dump_stream_from_url))
+            .await
+            .transpose()?;
     app_state
         .namespaces
-        .create(namespace.into(), Some(dump))
+        .create(namespace.into(), maybe_dump)
         .await?;
     Ok(())
 }
 
-async fn handle_create_namespace<F: MakeNamespace>(
-    State(app_state): State<Arc<AppState<F>>>,
-    Path(namespace): Path<String>,
-) -> Result<(), crate::error::Error> {
-    app_state.namespaces.create(namespace.into(), None).await?;
-    Ok(())
+async fn dump_stream_from_url(url: &Url) -> Result<DumpStream, LoadDumpError> {
+    match url.scheme() {
+        "http" => {
+            let client = hyper::client::Client::new();
+            let uri = url
+                .as_str()
+                .parse()
+                .map_err(|_| LoadDumpError::InvalidDumpUrl)?;
+            let resp = client.get(uri).await?;
+            let body = resp
+                .into_body()
+                .map_err(|e| std::io::Error::new(ErrorKind::Other, e));
+            Ok(Box::new(body))
+        }
+        "file" => {
+            let path = url
+                .to_file_path()
+                .map_err(|_| LoadDumpError::InvalidDumpUrl)?;
+            if !path.is_absolute() {
+                return Err(LoadDumpError::DumpFilePathNotAbsolute);
+            }
+
+            if !path.try_exists()? {
+                return Err(LoadDumpError::DumpFileDoesntExist);
+            }
+
+            let f = tokio::fs::File::open(path).await?;
+
+            Ok(Box::new(ReaderStream::new(f)))
+        }
+        _ => todo!("unsupported url scheme"),
```

Comment:
Should this respond with an error instead of a todo panic?

Reply:
oops yes